cpp/unreal_plugins/UrdfBot/Source/UrdfParser.h

```diff
@@ -0,0 +1,175 @@
+//
+// Copyright(c) 2022 Intel. Licensed under the MIT License <http://opensource.org/licenses/MIT>.
+//
+
+#pragma once
+
+#include <map>
+#include <vector>
+
+#include <Eigen/Dense>
+
+#include <CoreMinimal.h>
+
+class FXmlNode;
+
+struct UrdfJointDesc;
+
+enum class UrdfGeometryType
+{
+    Invalid  = 0,
+    Box      = 1,
+    Sphere   = 2,
+    Cylinder = 3,
+    Mesh     = 4,
+};
+
+enum class UrdfJointType
+{
+    Invalid    = 0,
+    Revolute   = 1,
+    Continuous = 2,
+    Prismatic  = 3,
+    Fixed      = 4,
+    Floating   = 5,
+    Planar     = 6
+};
+
+enum class CalibrationType
+{
+    Invalid = 0,
+    Rising  = 1,
+    Falling = 2
+};
+
+struct UrdfGeometryDesc
+{
+    UrdfGeometryType type_;
+    FVector size_ = FVector::ZeroVector; // required by Box, in [m]
+    float radius_ = 0.0f; // required by Sphere, Cylinder, in [m]
+    float length_ = 0.0f; // required by Cylinder, in [m]
+    std::string filename_; // required by Mesh
+    float scale_ ; // optional for Mesh
+};
+
+struct UrdfMaterialDesc
+{
+    std::string name_;
+
+    FVector4 color_;
+    std::string texture_;
+
+    bool is_reference_ = false;
+    UrdfMaterialDesc* material_desc_ = nullptr;
+};
+
+struct UrdfInertialDesc
+{
+    FTransform origin_ = FTransform::Identity;
+    float mass_ = 0.0f;
+    Eigen::Matrix3f inertia_;
+};
+
+struct UrdfVisualDesc
```

Comment:
```
struct UrdfVisualDesc
{
    std::string name_;

    FTransform origin_ = FTransform::Identity;
    UrdfGeometryDesc geometry_desc_;
    UrdfMaterialDesc material_desc_;
};
```

Reply:
fixed

Reply:
Based on your most recent comment, I think there should be some extra derived data here to account for the fact that a material node may or may not be present.

```
struct UrdfVisualDesc
{
     std::string name_;

    FTransform origin_ = FTransform::Identity;
    UrdfGeometryDesc geometry_desc_;
    UrdfMaterialDesc material_desc_;

    // derived data
    bool has_material = false;
};
```

Reply:
It is actually an interesting point. Unreal do support multiple material slot for single mesh and we could extend original urdf such that a visual node may have more than one material node.

Reply:
This is an interesting point. As we discussed in our meeting, let's consider this extension when we have a concrete need for it. But for now, let's assume there is 0 or 1.